# Part 0
Name: Oceana Zhu
NetID: oz2060

# Part 1

In [2]:
# Question 1
import numpy as np
import pandas as pd

Question 2:

In [3]:
#a
commute = [30,62,40,42,39]

In [4]:
#b
sum(commute)/len(commute)

42.6

On average, Charlie's commute is 42.6 minutes long. 

In [5]:
#c
commute_array = np.array(commute)
print(commute_array)

[30 62 40 42 39]


In [6]:
#d
commute_array_seconds = [i * 60 for i in commute_array]
print(commute_array_seconds)

[1800, 3720, 2400, 2520, 2340]


In [7]:
#d
commute_seconds = [i * 60 for i in commute]
print(commute_seconds)

[1800, 3720, 2400, 2520, 2340]


Multiplying the numpy array by 60 does still give you the same answer. 

In [52]:
#e
np.mean(commute_array)

42.6

On average, Charlie's commute is 42.6 minutes long. 

In [9]:
#f
np.max(commute_array_seconds) - np.mean(commute_array_seconds)

1164.0

Charlie's longest commute time was longer than the average commute time by 1164 seconds. 

In [10]:
#g
np.mean(commute_array_seconds) - np.min(commute_array_seconds)

756.0

Charlie's shortest commute time was shorter than his average commute time by 756 seconds. 

Question 3

In [11]:
fave_westvillage_restaurants = ['tacombi', 'grey dog', 'chloe', 'pommes frites', 'pho bar']
len(fave_westvillage_restaurants)

5

Question 4

In [12]:
#a
beef = 11.48
ques = 9.62
salestax = 0.08875

print((beef+ques)*(1+salestax))

22.972625000000004


In [13]:
#b
subtotal = beef+ques
tip = 3.15
print(tip/subtotal)

0.14928909952606634


Explanation of process: 
For part a: The sales tax is applied on the bill as a whole before the tax. This means that we need to find the sum of all the goods purchased (beef and ques) and apply a sales tax to it. You need to add 1 to the value of the sales tax before multiplication because you are paying the full value of the beef and ques along with the sales tax. 

For part b: The tip is applied tothe bill before the tax. This means that we need to find the sum of all the goods purchased (subtotal value). We are given the value of the tip, thus, we divide this value by the subtotal to figure out how much of the total the tip was. 

# Part 2

Question 5
Description of top50.csv: 
The top50 csv is a dataset that contains information about the top 50 songs. There are 13 variable and 50 songs analyzed. Instead of being presented in a table or chart, all of the values in this document are separated by commas. 

Question 6

In [14]:
my_first_dataframe = pd.read_csv('shared/datasets/top50.csv')
my_first_dataframe

TrackName        ArtistName  \
0                                            Señorita      Shawn Mendes   
1                                               China          Anuel AA   
2                       boyfriend (with Social House)     Ariana Grande   
3                     Beautiful People (feat. Khalid)        Ed Sheeran   
4                         Goodbyes (Feat. Young Thug)       Post Malone   
5                   I Don't Care (with Justin Bieber)        Ed Sheeran   
6                                              Ransom         Lil Tecca   
7                                   How Do You Sleep?         Sam Smith   
8                               Old Town Road - Remix         Lil Nas X   
9                                             bad guy     Billie Eilish   
10                                           Callaita         Bad Bunny   
11              Loco Contigo (feat. J. Balvin & Tyga)          DJ Snake   
12                                  Someone You Loved     Lewis Capaldi   
13                                 Otro Trago - Remix              Sech   
14           Money In The Grave (Drake ft. Rick Ross)             Drake   
15                          No Guidance (feat. Drake)       Chris Brown   
16                                         LA CANCIÓN          J Balvin   
17      Sunflower - Spider-Man: Into the Spider-Verse       Post Malone   
18                                             Lalala               Y2K   
19                                        Truth Hurts             Lizzo   
20                                Piece Of Your Heart            MEDUZA   
21                                             Panini         Lil Nas X   
22                               No Me Conoce - Remix       Jhay Cortez   
23                                    Soltera - Remix             Lunay   
24                       bad guy (with Justin Bieber)     Billie Eilish   
25                                If I Can't Have You      Shawn Mendes   
26                                       Dance Monkey       Tones and I   
27                                           It's You         Ali Gatie   
28                                          Con Calma      Daddy Yankee   
29                                      QUE PRETENDES          J Balvin   
30                                           Takeaway  The Chainsmokers   
31                                            7 rings     Ariana Grande   
32                                  0.958333333333333            Maluma   
33          The London (feat. J. Cole & Travis Scott)        Young Thug   
34                                  Never Really Over        Katy Perry   
35  Summer Days (feat. Macklemore & Patrick Stump ...     Martin Garrix   
36                                         Otro Trago              Sech   
37                     Antisocial (with Travis Scott)        Ed Sheeran   
38                                             Sucker    Jonas Brothers   
39  fuck, i'm lonely (with Anne-Marie) - from “13 ...              Lauv   
40                                        Higher Love              Kygo   
41                              You Need To Calm Down      Taylor Swift   
42                                            Shallow         Lady Gaga   
43                                               Talk            Khalid   
44                                         Con Altura           ROSALÍA   
45                                    One Thing Right        Marshmello   
46                                          Te Robaré         Nicky Jam   
47                                            Happier        Marshmello   
48                                      Call You Mine  The Chainsmokers   
49      Cross Me (feat. Chance the Rapper & PnB Rock)        Ed Sheeran   

               Genre  BeatsPerMinute  Energy  Danceability  LoudnessdB  \
0       canadian pop             117      55            76          -6   
1     reggaeton flow             105      81            79          -4

Kind of information contained in the dataframe: The dataframe contains information about the 50 top songs. It breaks down quantitative data about the songs, such as the beats per minute, but also qualitative data measured on a scale, such as danceability. Furthermore, there is also general information about the song, such as its name and artist. 

Question 7

In [15]:
print( my_first_dataframe.shape )
print( my_first_dataframe.columns )
print( my_first_dataframe.dtypes )

(50, 13)
Index(['TrackName', 'ArtistName', 'Genre', 'BeatsPerMinute', 'Energy',
       'Danceability', 'LoudnessdB', 'Liveness', 'Valence', 'Length',
       'Acousticness', 'Speechiness', 'Popularity'],
      dtype='object')
TrackName         object
ArtistName        object
Genre             object
BeatsPerMinute     int64
Energy             int64
Danceability       int64
LoudnessdB         int64
Liveness           int64
Valence            int64
Length             int64
Acousticness       int64
Speechiness        int64
Popularity         int64
dtype: object


What I learned about the dataframe: I learned that this dataframe has 13 columns and 50 observations or rows. I also learned the differnt names of the columns and the type of data that is recorded by each variable. 

Question 8

In [16]:
my_first_dataframe.head()

TrackName     ArtistName           Genre  \
0                         Señorita   Shawn Mendes    canadian pop   
1                            China       Anuel AA  reggaeton flow   
2    boyfriend (with Social House)  Ariana Grande       dance pop   
3  Beautiful People (feat. Khalid)     Ed Sheeran             pop   
4      Goodbyes (Feat. Young Thug)    Post Malone         dfw rap   

   BeatsPerMinute  Energy  Danceability  LoudnessdB  Liveness  Valence  \
0             117      55            76          -6         8       75   
1             105      81            79          -4         8       61   
2             190      80            40          -4        16       70   
3              93      65            64          -8         8       55   
4             150      65            58          -4        11       18   

   Length  Acousticness  Speechiness  Popularity  
0     191             4            3          79  
1     302             8            9          92  
2     186            12           46          85  
3     198            12           19          86  
4     175            45            7          94

In [17]:
my_first_dataframe.tail()

TrackName        ArtistName    Genre  \
45                                One Thing Right        Marshmello  brostep   
46                                      Te Robaré         Nicky Jam    latin   
47                                        Happier        Marshmello  brostep   
48                                  Call You Mine  The Chainsmokers      edm   
49  Cross Me (feat. Chance the Rapper & PnB Rock)        Ed Sheeran      pop   

    BeatsPerMinute  Energy  Danceability  LoudnessdB  Liveness  Valence  \
45              88      62            66          -2        58       44   
46             176      75            67          -4         8       80   
47             100      79            69          -3        17       67   
48             104      70            59          -6        41       50   
49              95      79            75          -6         7       61   

    Length  Acousticness  Speechiness  Popularity  
45     182             7            5          88  
46     202            24            6          88  
47     214            19            5          88  
48     218            23            3          88  
49     206            21           12          82

In [18]:
my_first_dataframe.head(3)

TrackName     ArtistName           Genre  \
0                       Señorita   Shawn Mendes    canadian pop   
1                          China       Anuel AA  reggaeton flow   
2  boyfriend (with Social House)  Ariana Grande       dance pop   

   BeatsPerMinute  Energy  Danceability  LoudnessdB  Liveness  Valence  \
0             117      55            76          -6         8       75   
1             105      81            79          -4         8       61   
2             190      80            40          -4        16       70   

   Length  Acousticness  Speechiness  Popularity  
0     191             4            3          79  
1     302             8            9          92  
2     186            12           46          85

In [19]:
my_first_dataframe.head(2)

TrackName    ArtistName           Genre  BeatsPerMinute  Energy  \
0  Señorita  Shawn Mendes    canadian pop             117      55   
1     China      Anuel AA  reggaeton flow             105      81   

   Danceability  LoudnessdB  Liveness  Valence  Length  Acousticness  \
0            76          -6         8       75     191             4   
1            79          -4         8       61     302             8   

   Speechiness  Popularity  
0            3          79  
1            9          92

In [20]:
my_first_dataframe.tail(2)

TrackName        ArtistName Genre  \
48                                  Call You Mine  The Chainsmokers   edm   
49  Cross Me (feat. Chance the Rapper & PnB Rock)        Ed Sheeran   pop   

    BeatsPerMinute  Energy  Danceability  LoudnessdB  Liveness  Valence  \
48             104      70            59          -6        41       50   
49              95      79            75          -6         7       61   

    Length  Acousticness  Speechiness  Popularity  
48     218            23            3          88  
49     206            21           12          82

In [21]:
my_first_dataframe.tail(3)

TrackName        ArtistName    Genre  \
47                                        Happier        Marshmello  brostep   
48                                  Call You Mine  The Chainsmokers      edm   
49  Cross Me (feat. Chance the Rapper & PnB Rock)        Ed Sheeran      pop   

    BeatsPerMinute  Energy  Danceability  LoudnessdB  Liveness  Valence  \
47             100      79            69          -3        17       67   
48             104      70            59          -6        41       50   
49              95      79            75          -6         7       61   

    Length  Acousticness  Speechiness  Popularity  
47     214            19            5          88  
48     218            23            3          88  
49     206            21           12          82

In [22]:
my_first_dataframe.sample(5)

TrackName        ArtistName        Genre  \
15                  No Guidance (feat. Drake)       Chris Brown    dance pop   
12                          Someone You Loved     Lewis Capaldi          pop   
33  The London (feat. J. Cole & Travis Scott)        Young Thug  atl hip hop   
30                                   Takeaway  The Chainsmokers          edm   
9                                     bad guy     Billie Eilish   electropop   

    BeatsPerMinute  Energy  Danceability  LoudnessdB  Liveness  Valence  \
15              93      45            70          -7        16       14   
12             110      41            50          -6        11       45   
33              98      59            80          -7        13       18   
30              85      51            29          -8        10       36   
9              135      43            70         -11        10       56   

    Length  Acousticness  Speechiness  Popularity  
15     261            12           15          82  
12     182            75            3          88  
33     200             2           15          89  
30     210            12            4          84  
9      194            33           38          95

In [23]:
my_first_dataframe.sample(3)

TrackName     ArtistName      Genre  BeatsPerMinute  Energy  \
12      Someone You Loved  Lewis Capaldi        pop             110      41   
41  You Need To Calm Down   Taylor Swift  dance pop              85      68   
34      Never Really Over     Katy Perry  dance pop             100      88   

    Danceability  LoudnessdB  Liveness  Valence  Length  Acousticness  \
12            50          -6        11       45     182            75   
41            77          -6         7       73     171             1   
34            77          -5        32       39     224            19   

    Speechiness  Popularity  
12            3          88  
41            5          90  
34            6          89

What was done by each of the commands: The head command extracted the first n rows from a dataset. The tail command extracted the last n rows from a dataset. The sample command extracted a random sample of n rows from a dataset. 

# Part 3

Question 9

In [24]:
bronx_pop_density = pd.read_csv('https://bronx.lehman.cuny.edu/api/views/2q3n-mzjq/rows.csv?accessType=DOWNLOAD')                              

In [25]:
bronx_pop_density.head(2)

Zipcode County and Neighborhood  Total Area Sq.Miles      \
0    10452       Bronx High Bridge                     0.98   
1    10453        Bronx Morris Hts                     0.92   

   Total Area Sq.Yard  Pop.Density per Sq Mile  Pop.Density per Sq Yard  \
0           3035648.0               76909.1837                   0.0248   
1           2849792.0               85118.4783                   0.0275   

   Total Population 2010  Pop. Age 65 and Up  Pop. Under 18  Households  ...  \
0                  75371                5237          22527       24796  ...   
1                  78309                5056          23464       25307  ...   

   Age 65 and Up: % Live Alone  Age 65 and Up: % Live in Fams  \
0                         30.3                           64.6   
1                         27.5                           67.1   

   Age 65 and Up: % in Gp.Quarters  % Kids Live w/Ma+Pa  \
0                              0.1                 32.3   
1                              0.4                 27.3   

   % Kids Live w/Ma Only  % Kids Live w/Grandparents  \
0                   41.5                        10.8   
1                   46.3                        11.8   

   % Kids Live in Group Quarters  Housing Units  % Rental Units  \
0                            1.6          25866            92.8   
1                            1.6          27018            89.4   

                        Location 1  
0       (40.83724, -73.9211399997)  
1  (40.8522899998, -73.9109799997)  

[2 rows x 50 columns]

In [26]:
bronx_pop_density.tail(2)

Zipcode County and Neighborhood  Total Area Sq.Miles      \
23    10475         Bronx Coop City                     1.82   
24    10451           Bronx Melrose                     1.02   

    Total Area Sq.Yard  Pop.Density per Sq Mile  Pop.Density per Sq Yard  \
23           5637632.0               22489.5604                   0.0073   
24           3159552.0               44816.6667                   0.0145   

    Total Population 2010  Pop. Age 65 and Up  Pop. Under 18  Households  ...  \
23                  40931                8760           7513       18534  ...   
24                  45713                5453          12444       17106  ...   

    Age 65 and Up: % Live Alone  Age 65 and Up: % Live in Fams  \
23                         37.0                           52.0   
24                         42.8                           52.7   

    Age 65 and Up: % in Gp.Quarters  % Kids Live w/Ma+Pa  \
23                              8.5                 39.2   
24                              0.8                 28.2   

    % Kids Live w/Ma Only  % Kids Live w/Grandparents  \
23                   32.9                        14.2   
24                   45.7                        14.1   

    % Kids Live in Group Quarters  Housing Units  % Rental Units  \
23                            1.6          18911            51.3   
24                            0.2          17779            82.7   

                         Location 1  
23  (40.8728999996, -73.8270299999)  
24  (40.8221799998, -73.9209600004)  

[2 rows x 50 columns]

Describe what information is contained in this data frame: This data frame contains information about the the demographics of the population in the Bronx, broken down by neighborhood. The data frame contains information about race, age, family size, housing, and general information about the area. 

Question 10:
How many rows and columns are there in this dataset? There are 25 rows and 50 columns. 
State three of the column names and their corresponding data types
Three column names are Zipcode, Country and Neighborhood, and Total Area Sq.Miles. Their corresponding data types are integer, text, and floating point number. 

In [27]:
print( bronx_pop_density.shape )
print( bronx_pop_density.dtypes )

(25, 50)
Zipcode                              int64
County and Neighborhood             object
Total Area Sq.Miles                float64
Total Area Sq.Yard                 float64
Pop.Density per Sq Mile            float64
Pop.Density per Sq Yard            float64
Total Population 2010                int64
Pop. Age 65 and Up                   int64
Pop. Under 18                        int64
Households                           int64
Families                             int64
Non-Hispanic Whites                  int64
Non-Hispanic Blacks                  int64
Non-Hispanic Asians                  int64
Hispanics                            int64
% Non-Hisp. Whites                 float64
% Non-Hisp. Blacks                 float64
% Non-Hisp.Asians                  float64
% Hispanics                        float64
% Two or More Races                float64
Median Age: Blacks                 float64
Median Age: Asians                 float64
Median Age: Hispanics              float64
Me

Question 11
a - Create a new column called fraction_under18, which contains the fraction of each zipcode's population who are under 18
b - Which zipcode has the highest fraction of residents under 18? Is this zipcode a lot “younger” than other zipcodes in the Bronx? Use a dataframe method to sort the rows.

In [32]:
#a
bronx_pop_density['fraction_under18'] = bronx_pop_density['Pop. Under 18'] / bronx_pop_density['Total Population 2010']
#b
bronx_pop_density.sort_values (by ='fraction_under18', axis=0, ascending=False)

Zipcode County and Neighborhood  Total Area Sq.Miles      \
5     10457           Bronx Tremont                     1.06   
4     10456        Bronx Morrisania                     1.02   
7     10459  Bronx Boulevd-Longwood                     0.82   
8     10460        Bronx West Farms                     1.32   
2     10454        Bronx Mott Haven                     1.04   
22    10474       Bronx Hunts Point                     1.53   
1     10453        Bronx Morris Hts                     0.92   
3     10455               Bronx Hub                     0.71   
0     10452       Bronx High Bridge                     0.98   
6     10458           Bronx Fordham                     1.01   
20    10472          Bronx Bruckner                     1.05   
16    10468            Bronx Jerome                     1.09   
24    10451           Bronx Melrose                     1.02   
21    10473         Bronx Soundview                     2.19   
15    10467    Bronx Williamsbridge                     2.33   
14    10466         Bronx Wakefield                     2.02   
17    10469        Bronx Baychester                     2.48   
10    10462       Bronx Parkchester                     1.48   
11    10463       Bronx Kingsbridge                     1.58   
18    10470          Bronx Woodlawn                     1.43   
13    10465       Bronx Throgs Neck                     3.43   
9     10461       Bronx Westchester                     2.40   
23    10475         Bronx Coop City                     1.82   
12    10464       Bronx City Island                     3.59   
19    10471         Bronx Riverdale                     2.59   

    Total Area Sq.Yard  Pop.Density per Sq Mile  Pop.Density per Sq Yard  \
5            3283456.0               66505.6604                   0.0215   
4            3159552.0               84850.0000                   0.0274   
7            2540032.0               57692.6829                   0.0186   
8            4088832.0               43417.4242                   0.0140   
2            3221504.0               35900.9615                   0.0116   
22           4739328.0                8026.7974                   0.0026   
1            2849792.0               85118.4783                   0.0275   
3            2199296.0               55866.1972                   0.0180   
0            3035648.0               76909.1837                   0.0248   
6            3128576.0               78704.9505                   0.0254   
20           3252480.0               63198.0952                   0.0204   
16           3376384.0               69819.2661                   0.0225   
24           3159552.0               44816.6667                   0.0145   
21           6783744.0               26721.0046                   0.0086   
15           7217408.0               41656.6524                   0.0134   
14           6257152.0               33570.7921                   0.0108   
17           7682048.0               26867.3387                   0.0087   
10           4584448.0               51205.4054                   0.0165   
11           4894208.0               37662.6582                   0.0122   
18           4429568.0               10694.4056                   0.0035   
13          10624768.0               12311.9534                   0.0040   
9            7434240.0               21042.5000                   0.0068   
23           5637632.0               22489.5604                   0.0073   
12          11120384.0                1262.9526                   0.0004   
19           8022784.0                8850.1931                   0.0029   

    Total Population 2010  Pop. Age 65 and Up  Pop. Under 18  Households  ...  \
5                   70496                5304          21943       22873  ...   
4                   86547                6713          26930       28122  ...   
7                   47308                3957          14616       15398  ...   
8                   57311             

10457 is the zip code with the highest fraction of residents under 18. I would say that this zipcode is "younger" than other zipcodes because it has the greatest percentage of its population under the age of 18. 

Question 12
Please formulate a question that can be answered by creating a new column in the above data frame. (The process would be similar to the above exercise, but you are asked to come up with your own question!) Please include the python codes/commands as well as a text summary of both your question and the answer

Another question that can be answered is: What zipcode has the greatest percentage of residents above the age of 65. 

In [34]:
bronx_pop_density['fraction_above65'] = bronx_pop_density['Pop. Age 65 and Up'] / bronx_pop_density['Total Population 2010']
bronx_pop_density.sort_values (by ='fraction_above65', axis=0, ascending=False)

Zipcode County and Neighborhood  Total Area Sq.Miles      \
19    10471         Bronx Riverdale                     2.59   
23    10475         Bronx Coop City                     1.82   
12    10464       Bronx City Island                     3.59   
11    10463       Bronx Kingsbridge                     1.58   
13    10465       Bronx Throgs Neck                     3.43   
9     10461       Bronx Westchester                     2.40   
18    10470          Bronx Woodlawn                     1.43   
17    10469        Bronx Baychester                     2.48   
24    10451           Bronx Melrose                     1.02   
21    10473         Bronx Soundview                     2.19   
14    10466         Bronx Wakefield                     2.02   
10    10462       Bronx Parkchester                     1.48   
15    10467    Bronx Williamsbridge                     2.33   
20    10472          Bronx Bruckner                     1.05   
2     10454        Bronx Mott Haven                     1.04   
3     10455               Bronx Hub                     0.71   
8     10460        Bronx West Farms                     1.32   
16    10468            Bronx Jerome                     1.09   
7     10459  Bronx Boulevd-Longwood                     0.82   
4     10456        Bronx Morrisania                     1.02   
5     10457           Bronx Tremont                     1.06   
0     10452       Bronx High Bridge                     0.98   
6     10458           Bronx Fordham                     1.01   
1     10453        Bronx Morris Hts                     0.92   
22    10474       Bronx Hunts Point                     1.53   

    Total Area Sq.Yard  Pop.Density per Sq Mile  Pop.Density per Sq Yard  \
19           8022784.0                8850.1931                   0.0029   
23           5637632.0               22489.5604                   0.0073   
12          11120384.0                1262.9526                   0.0004   
11           4894208.0               37662.6582                   0.0122   
13          10624768.0               12311.9534                   0.0040   
9            7434240.0               21042.5000                   0.0068   
18           4429568.0               10694.4056                   0.0035   
17           7682048.0               26867.3387                   0.0087   
24           3159552.0               44816.6667                   0.0145   
21           6783744.0               26721.0046                   0.0086   
14           6257152.0               33570.7921                   0.0108   
10           4584448.0               51205.4054                   0.0165   
15           7217408.0               41656.6524                   0.0134   
20           3252480.0               63198.0952                   0.0204   
2            3221504.0               35900.9615                   0.0116   
3            2199296.0               55866.1972                   0.0180   
8            4088832.0               43417.4242                   0.0140   
16           3376384.0               69819.2661                   0.0225   
7            2540032.0               57692.6829                   0.0186   
4            3159552.0               84850.0000                   0.0274   
5            3283456.0               66505.6604                   0.0215   
0            3035648.0               76909.1837                   0.0248   
6            3128576.0               78704.9505                   0.0254   
1            2849792.0               85118.4783                   0.0275   
22           4739328.0                8026.7974                   0.0026   

    Total Population 2010  Pop. Age 65 and Up  Pop. Under 18  Households  ...  \
19                  22922                5376           3700        9305  ...   
23                  40931                8760           7513       18534  ...   
12                   4534                 917            787        2089  ...   
11                  59507             

The answer to the question of which zip code has the oldest population can be answered by creating a new column that calculates the fraction of the total population of that area that is above the age of 65. The answer to this question is 10471 because they had the greatest fraction of people above the age of 65. 

Question 13
Suppose that we are interested in finding out the percentage of the population in Bronx Kingsbridge who stated that they belong to two or more races. Type a python code/command that extracts the above number from the dataframe. Please very briefly describe how you think about this question.

In [36]:
print(bronx_pop_density[bronx_pop_density['County and Neighborhood']=='Bronx Kingsbridge'].index.values)

[11]


In [37]:
bronx_pop_density.loc[11,"% Two or More Races"]

5.0

I thought about this question by thinking about the fact that I have to locate a value given the fact that we are interested in the county of Bronx Kingsbridge. Thus, I found the index of Bronx Kingsbridge and located the value of "% Two or More Races" at that index value for Bronx Kingsbridge. 

Question 14
Pick two columns (your choice) and simultaneously sort bronx pop density based on values in these two columns; one in ascending order and the other in descending order. In the “writing” portion of this question, briefly summarize what you did (that is, state which two columns you pick, etc.). Does the resulting sorted data frame match what you expected to see?

In [49]:
bronx_pop_density.sort_values(by=['Housing Units', '% Rental Units'], ascending=[True, False])

Zipcode County and Neighborhood  Total Area Sq.Miles      \
12    10464       Bronx City Island                     3.59   
22    10474       Bronx Hunts Point                     1.53   
18    10470          Bronx Woodlawn                     1.43   
19    10471         Bronx Riverdale                     2.59   
2     10454        Bronx Mott Haven                     1.04   
3     10455               Bronx Hub                     0.71   
7     10459  Bronx Boulevd-Longwood                     0.82   
13    10465       Bronx Throgs Neck                     3.43   
24    10451           Bronx Melrose                     1.02   
23    10475         Bronx Coop City                     1.82   
8     10460        Bronx West Farms                     1.32   
9     10461       Bronx Westchester                     2.40   
21    10473         Bronx Soundview                     2.19   
20    10472          Bronx Bruckner                     1.05   
17    10469        Bronx Baychester                     2.48   
5     10457           Bronx Tremont                     1.06   
14    10466         Bronx Wakefield                     2.02   
0     10452       Bronx High Bridge                     0.98   
16    10468            Bronx Jerome                     1.09   
1     10453        Bronx Morris Hts                     0.92   
6     10458           Bronx Fordham                     1.01   
4     10456        Bronx Morrisania                     1.02   
11    10463       Bronx Kingsbridge                     1.58   
10    10462       Bronx Parkchester                     1.48   
15    10467    Bronx Williamsbridge                     2.33   

    Total Area Sq.Yard  Pop.Density per Sq Mile  Pop.Density per Sq Yard  \
12          11120384.0                1262.9526                   0.0004   
22           4739328.0                8026.7974                   0.0026   
18           4429568.0               10694.4056                   0.0035   
19           8022784.0                8850.1931                   0.0029   
2            3221504.0               35900.9615                   0.0116   
3            2199296.0               55866.1972                   0.0180   
7            2540032.0               57692.6829                   0.0186   
13          10624768.0               12311.9534                   0.0040   
24           3159552.0               44816.6667                   0.0145   
23           5637632.0               22489.5604                   0.0073   
8            4088832.0               43417.4242                   0.0140   
9            7434240.0               21042.5000                   0.0068   
21           6783744.0               26721.0046                   0.0086   
20           3252480.0               63198.0952                   0.0204   
17           7682048.0               26867.3387                   0.0087   
5            3283456.0               66505.6604                   0.0215   
14           6257152.0               33570.7921                   0.0108   
0            3035648.0               76909.1837                   0.0248   
16           3376384.0               69819.2661                   0.0225   
1            2849792.0               85118.4783                   0.0275   
6            3128576.0               78704.9505                   0.0254   
4            3159552.0               84850.0000                   0.0274   
11           4894208.0               37662.6582                   0.0122   
10           4584448.0               51205.4054                   0.0165   
15           7217408.0               41656.6524                   0.0134   

    Total Population 2010  Pop. Age 65 and Up  Pop. Under 18  Households  ...  \
12                   4534                 917            787        2089  ...   
22                  12281                 750           3684        3633  ...   
18                  15293                2196           3239        6390  ...   
19                  22922             

I sorted Housing Units in descending order and % Rental Units by ascending order. The dataframe almost matches what I expected to see because I was unsure how the second column would be sorted but thought that its sorting would be limited by the sorting of the first column. As such, I observed that while Housing Units were ordered from lowest to highest, % Rental Units wasn't perfectly ordered highest to lowest because it was constrained by the ordering of the Housing Units column. 

Question 15
Create a data frame called bronx pop density rental filtered containing only rows where the percentage of rental units are 70 or above. Then, determine how many percent of all neighborhoods/zipcodes in the Bronx satisfy the above criteria (70 percent or more of its housing unit are rentals). Please briefly describe how you approach this question.

In [50]:
bronx_pop_density_rental_filtered = bronx_pop_density[bronx_pop_density["% Rental Units"] >= 70]
print(bronx_pop_density_rental_filtered)

    Zipcode County and Neighborhood  Total Area Sq.Miles      \
0     10452       Bronx High Bridge                     0.98   
1     10453        Bronx Morris Hts                     0.92   
2     10454        Bronx Mott Haven                     1.04   
3     10455               Bronx Hub                     0.71   
4     10456        Bronx Morrisania                     1.02   
5     10457           Bronx Tremont                     1.06   
6     10458           Bronx Fordham                     1.01   
7     10459  Bronx Boulevd-Longwood                     0.82   
8     10460        Bronx West Farms                     1.32   
10    10462       Bronx Parkchester                     1.48   
15    10467    Bronx Williamsbridge                     2.33   
16    10468            Bronx Jerome                     1.09   
20    10472          Bronx Bruckner                     1.05   
21    10473         Bronx Soundview                     2.19   
22    10474       Bronx Hunts Point     

In [51]:
len(bronx_pop_density_rental_filtered.index)/len(bronx_pop_density.index)

0.64

64% of all neighborhoods/zipcodes satisfy the criteria of having 70% or more of its housing units be rentals. I approached the question by understanding that I had to create a dataframe that filtered out all of the entries that had % Rental Units lower than 70. Then I understood that the total number of neighborhoods would be equal to the number of rows. Thus, I used the index function on the old and new filtered datasets. By dividing the two index values, I would get the percent. 

# Part 4

Question 16
Indicate one question/part in this assignment that you find helpful in helping you understand an idea or method introduced in class. This could be a question/task that you knew right away how to do; one for which you kept encountering bugs/errors; or one that you needed to think a while or discuss with others first to figure out. (State which question/part and how it helps you with which idea/method.)

One question/part in this assignment that I found helpful in understanding an idea introduced in class were Question 11 and 12 because they helped me practice filtering datasets. While the coding portion of it wasn't challenging to figure out, it was helpful to be able to see how the dataset was changed after the filter was applied. 

Question 17
Write one thing that you find challenging or interesting in this assignment.

One thing that I found interesting in this assignment was the task in Question 14 where we had to filter the dataset by multiple columns. I found this interesting because it emulated the type of data analysis work that we would need to do in the real-world where a dataset would need to be analyzed through a variety of lenses. 